File: clients/javascript/src/resources/topics.ts
import type { HieroProxyClient } from "../client";
import type {
  TopicCreatedResponse,
  TopicResponse,
  TopicKeyRotationResponse,
  TopicUpdatedResponse,
  TopicMessageResponse,
  SubmitMessageRequest,
  SubmitBinaryMessageRequest,
  CreateTopicWithAdminKeyRequest,
  CreatePrivateTopicWithAdminKeyRequest,
  UpdateTopicMemoRequest,
  UpdateTopicKeyRequest,
  UpdateTopicRequest,
  DeleteTopicRequest,
} from "./tokens-topics-types";
import type { SuccessResponse } from "./types";

/** Validates that an ID is non-empty and URL-encodes it. */
function requireId(value: string, paramName: string): string {
  if (!value || typeof value !== "string" || !value.trim()) {
    throw new Error(`${paramName} is required and must be a non-empty string`);
  }
  return encodeURIComponent(value.trim());
}

/**
 * Topics resource — create, query, update, delete topics; submit and read messages.
 */
export class TopicsResource {
  constructor(private readonly client: HieroProxyClient) {}

  // ─── Creation ─────────────────────────────────────────────────────────────

  /**
   * Create a public topic (operator as admin key).
   */
  async create(memo?: string): Promise<TopicCreatedResponse> {
    const body = memo ? { memo } : undefined;
    return this.client.post("/api/v1/topics", body);
  }

  /**
   * Create a private topic (generates submit key, operator as admin key).
   */
  async createPrivate(memo?: string): Promise<TopicCreatedResponse> {
    const body = memo ? { memo } : undefined;
    return this.client.post("/api/v1/topics/private", body);
  }

  /**
   * Create a public topic with a custom admin key.
   */
  async createWithAdminKey(adminPrivateKey: string, memo?: string): Promise<TopicCreatedResponse> {
    const body: CreateTopicWithAdminKeyRequest = { adminPrivateKey, memo };
    return this.client.post("/api/v1/topics/with-admin-key", body);
  }

  /**
   * Create a private topic with a custom admin key (generates submit key).
   */
  async createPrivateWithAdminKey(adminPrivateKey: string, memo?: string): Promise<TopicCreatedResponse> {
    const body: CreatePrivateTopicWithAdminKeyRequest = { adminPrivateKey, memo };
    return this.client.post("/api/v1/topics/private/with-admin-key", body);
  }

  // ─── Queries ──────────────────────────────────────────────────────────────

  /**
   * Get topic info from the mirror node.
   */
  async getInfo(topicId: string): Promise<TopicResponse> {
    const id = requireId(topicId, "topicId");
    return this.client.get(`/api/v1/topics/${id}`);
  }

  // ─── Updates ──────────────────────────────────────────────────────────────

  /**
   * Update topic memo.
   */
  async updateMemo(topicId: string, adminPrivateKey: string, memo: string): Promise<SuccessResponse> {
    const id = requireId(topicId, "topicId");
    const body: UpdateTopicMemoRequest = { adminPrivateKey, memo };
    return this.client.put(`/api/v1/topics/${id}/memo`, body);
  }

  /**
   * Rotate the admin key of a topic. Returns the new key pair.
   */
  async rotateAdminKey(topicId: string, adminPrivateKey: string): Promise<TopicKeyRotationResponse> {
    const id = requireId(topicId, "topicId");
    const body: UpdateTopicKeyRequest = { adminPrivateKey };
    return this.client.put(`/api/v1/topics/${id}/admin-key`, body);
  }

  /**
   * Rotate the submit key of a private topic. Returns the new key pair.
   */
  async rotateSubmitKey(topicId: string, adminPrivateKey: string): Promise<TopicKeyRotationResponse> {
    const id = requireId(topicId, "topicId");
    const body: UpdateTopicKeyRequest = { adminPrivateKey };
    return this.client.put(`/api/v1/topics/${id}/submit-key`, body);
  }

  /**
   * Atomically update admin key, submit key, and memo.
   */
  async update(topicId: string, currentAdminPrivateKey: string, memo: string): Promise<TopicUpdatedResponse> {
    const id = requireId(topicId, "topicId");
    const body: UpdateTopicRequest = { currentAdminPrivateKey, memo };
    return this.client.put(`/api/v1/topics/${id}`, body);
  }

  // ─── Deletion ─────────────────────────────────────────────────────────────

  /**
   * Delete a topic.
   */
  async delete(topicId: string, adminPrivateKey: string): Promise<SuccessResponse> {
    const id = requireId(topicId, "topicId");
    const body: DeleteTopicRequest = { adminPrivateKey };
    return this.client.delete(`/api/v1/topics/${id}`, body);
  }

  // ─── Messages ─────────────────────────────────────────────────────────────

  /**
   * Submit a text message to a topic.
   * For private topics, provide the submitKey.
   */
  async submitMessage(topicId: string, message: string, submitKey?: string): Promise<SuccessResponse> {
    const id = requireId(topicId, "topicId");
    const body: SubmitMessageRequest = { message, submitKey };
    return this.client.post(`/api/v1/topics/${id}/messages`, body);
  }

  /**
   * Submit a binary message (Base64-encoded) to a topic.
   * For private topics, provide the submitKey.
   */
  async submitBinaryMessage(topicId: string, messageBase64: string, submitKey?: string): Promise<SuccessResponse> {
    const id = requireId(topicId, "topicId");
    const body: SubmitBinaryMessageRequest = { messageBase64, submitKey };
    return this.client.post(`/api/v1/topics/${id}/messages/binary`, body);
  }

  /**
   * Get all messages for a topic.
   */
  async getMessages(topicId: string): Promise<TopicMessageResponse[]> {
    const id = requireId(topicId, "topicId");
    return this.client.get(`/api/v1/topics/${id}/messages`);
  }

  /**
   * Get a specific message by sequence number.
   */
  async getMessage(topicId: string, sequenceNumber: number): Promise<TopicMessageResponse> {
    const id = requireId(topicId, "topicId");
    return this.client.get(`/api/v1/topics/${id}/messages/${sequenceNumber}`);
  }
}

File: clients/python/src/hiero_proxy/resources/topics.py
"""Topics resource."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from hiero_proxy.resources.validation import require_id

if TYPE_CHECKING:
    from hiero_proxy.client import AsyncHieroProxyClient, HieroProxyClient


class TopicsResource:
    """Topic operations — create, submit messages, query, update, delete."""

    def __init__(self, client: HieroProxyClient | AsyncHieroProxyClient) -> None:
        self._client = client

    def create(self, memo: str = "") -> Any:
        return self._client.post("/api/v1/topics", json={"memo": memo} if memo else None)

    def create_private(self, memo: str = "") -> Any:
        return self._client.post("/api/v1/topics/private", json={"memo": memo} if memo else None)

    def create_with_admin_key(self, memo: str = "") -> Any:
        return self._client.post("/api/v1/topics/admin", json={"memo": memo} if memo else None)

    def create_private_with_admin_key(self, memo: str = "") -> Any:
        return self._client.post("/api/v1/topics/private/admin", json={"memo": memo} if memo else None)

    def get_info(self, topic_id: str) -> Any:
        tid = require_id(topic_id, "topic_id")
        return self._client.get(f"/api/v1/topics/{tid}")

    def update_memo(self, topic_id: str, memo: str) -> Any:
        tid = require_id(topic_id, "topic_id")
        return self._client.put(f"/api/v1/topics/{tid}/memo", json={"memo": memo})

    def rotate_admin_key(self, topic_id: str) -> Any:
        tid = require_id(topic_id, "topic_id")
        return self._client.put(f"/api/v1/topics/{tid}/keys/admin")

    def rotate_submit_key(self, topic_id: str) -> Any:
        tid = require_id(topic_id, "topic_id")
        return self._client.put(f"/api/v1/topics/{tid}/keys/submit")

    def update(self, topic_id: str, memo: str) -> Any:
        tid = require_id(topic_id, "topic_id")
        return self._client.put(f"/api/v1/topics/{tid}", json={"memo": memo})

    def delete(self, topic_id: str) -> Any:
        tid = require_id(topic_id, "topic_id")
        return self._client.delete(f"/api/v1/topics/{tid}")

    def submit_message(self, topic_id: str, message: str) -> Any:
        tid = require_id(topic_id, "topic_id")
        return self._client.post(f"/api/v1/topics/{tid}/message", json={"message": message})

    def submit_binary_message(self, topic_id: str, message: str, private_key: str) -> Any:
        tid = require_id(topic_id, "topic_id")
        return self._client.post(f"/api/v1/topics/{tid}/message/binary", json={
            "message": message, "privateKey": private_key,
        })

    def get_messages(self, topic_id: str) -> Any:
        tid = require_id(topic_id, "topic_id")
        return self._client.get(f"/api/v1/topics/{tid}/messages")

    def get_message(self, topic_id: str, sequence_number: int) -> Any:
        tid = require_id(topic_id, "topic_id")
        return self._client.get(f"/api/v1/topics/{tid}/messages/{sequence_number}")
